packages/server/lib/StudioLifecycleManager.ts

```diff
@@ -0,0 +1,151 @@
+   * Also registers this instance in the data context.
+   * @param projectId The project ID
+   * @param cloudDataSource The cloud data source
+   * @param cfg The project configuration
+   * @param debugData Debug data for the configuration
+   * @param ctx Data context to register this instance with
+   */
+  initializeStudioManager ({
+    projectId,
+    cloudDataSource,
+    cfg,
+    debugData,
+    ctx,
+  }: {
+    projectId?: string
+    cloudDataSource: CloudDataSource
+    cfg: Cfg
+    debugData: any
+    ctx: DataContext
+  }): void {
+    debug('Initializing studio manager')
+
+    const studioManagerPromise = getAndInitializeStudioManager({
+      projectId,
+      cloudDataSource,
+    }).then(async (studioManager) => {
+      if (studioManager.status === 'ENABLED') {
+        debug('Studio manager enabled, setting up protocol')
+        const protocolManager = new ProtocolManager()
+        const protocolUrl = routes.apiRoutes.captureProtocolCurrent()
+        const script = await api.getCaptureProtocolScript(protocolUrl)
+
+        await protocolManager.prepareProtocol(script, {
+          runId: 'studio',
+          projectId: cfg.projectId,
+          testingType: cfg.testingType,
+          cloudApi: {
+            url: routes.apiUrl,
+            retryWithBackoff: api.retryWithBackoff,
+            requestPromise: api.rp,
+          },
+          projectConfig: _.pick(cfg, ['devServerPublicPathRoute', 'port', 'proxyUrl', 'namespace']),
+          mountVersion: api.runnerCapabilities.protocolMountVersion,
+          debugData,
+          mode: 'studio',
+        })
+
+        studioManager.protocolManager = protocolManager
+        studioManager.isProtocolEnabled = true
+      }
+
+      return studioManager
+    }).catch((err) => {
+      debug('Error during studio manager setup: %o', err)
+
+      return null
+    })
+
+    this.setStudioManagerPromise(studioManagerPromise)
+
+    // Register this instance in the data context
+    ctx.update((data) => {
+      data.studioLifecycleManager = this
+    })
+  }
+
+  /**
+   * Check if studio manager is ready to be used
+   * @returns boolean indicating if studio manager promise exists
+   */
+  isStudioReady (): boolean {
+    return this.studioManagerPromise !== null
+  }
+
+  /**
+   * Get the studio manager if ready
+   * @returns The studio manager promise if ready, or null if not
+   */
+  getStudioIfReady (): Promise<StudioManager | null> | null {
+    return this.studioManagerPromise
+  }
+
+  async getStudio () {
+    if (!this.studioManagerPromise) {
+      throw new Error('Studio manager has not been initialized')
+    }
+
+    return this.studioManagerPromise
+  }
+
+  /**
+   * Register a listener that will be called when the studio is ready
+   * @param listener Function to call when studio is ready
+   * @returns Function to remove the listener
+   */
+  onStudioReady (listener: (studioManager: StudioManager) => void): () => void {
```

Comment:
Since this class is an `EventEmitter`, would it make more sense to just have the consumers register their own listeners?

Reply:
I'm not sure... my intention here was to control some of the logic around how listeners are set - like, we only want them to be called once, and only if studio manager actually was successfully initialized. I don't see the benefit of setting the listeners directly

Reply:
I might not be understanding what you mean

Reply:
We talked about this outside of this PR - I ended up re-working the lifecycle manager so that it doesn't use `EventEmitter`. It keeps track of all of the listeners that are registered and calls them when studio is ready. If a listener gets registered when studio is already ready, then it gets called immediately. Let me know what you think